src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ReconciliationService.java

```diff
@@ -0,0 +1,139 @@
+import com.microsoft.applicationinsights.core.dependencies.google.gson.GsonBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.camel.Exchange;
+import org.apache.commons.lang3.Validate;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.justice.digital.hmpps.keyworker.dto.OffenderLocationDto;
+import uk.gov.justice.digital.hmpps.keyworker.dto.PrisonerDetail;
+import uk.gov.justice.digital.hmpps.keyworker.dto.PrisonerIdentifier;
+import uk.gov.justice.digital.hmpps.keyworker.dto.SortOrder;
+import uk.gov.justice.digital.hmpps.keyworker.model.DeallocationReason;
+import uk.gov.justice.digital.hmpps.keyworker.model.OffenderKeyworker;
+import uk.gov.justice.digital.hmpps.keyworker.repository.OffenderKeyworkerRepository;
+
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
+@Slf4j
+public class ReconciliationService {
+
+    private final NomisService nomisService;
+    private final OffenderKeyworkerRepository offenderKeyworkerRepository;
+    private final TelemetryClient telemetryClient;
+    private final Gson gson;
+
+    public ReconciliationService(final NomisService nomisService,
+                                 final OffenderKeyworkerRepository offenderKeyworkerRepository,
+                                 final TelemetryClient telemetryClient) {
+        this.nomisService = nomisService;
+        this.offenderKeyworkerRepository = offenderKeyworkerRepository;
+        this.telemetryClient = telemetryClient;
+        gson = new GsonBuilder().create();
+    }
+
+    @Transactional
+    public void reconcileKeyWorkerAllocations(final String prisonId) {
+        Validate.notNull(prisonId,"prisonId");
+
+        // get all prisoners with an active key worker in prison
+        final var assignedPrisonersToKW = offenderKeyworkerRepository.findByActiveAndPrisonId(true, prisonId);
+        final var offenderNosInPrison = getOffenderNosInPrison(prisonId);
+
+        // find allocations no longer in this prison
+        var missingOffenders = assignedPrisonersToKW.stream()
+                .filter(kw -> !offenderNosInPrison.contains(kw.getOffenderNo()))
+                .collect(Collectors.toList());
+        log.info("There are {} missing prisoners from {}", missingOffenders.size(), prisonId);
+
+        final var notFoundOffenders = new AtomicInteger();
+        final var deallocatedOffenders = new AtomicInteger();
+        final var mergedRecords = new HashMap<String, String>();
+        missingOffenders.forEach( notFoundOffender -> nomisService.getPrisonerDetail(notFoundOffender.getOffenderNo()).ifPresentOrElse(
+                prisonerDetail -> deallocateIfMoved(prisonId, notFoundOffender, prisonerDetail, deallocatedOffenders),
+                () -> {
```

Comment:
probably tidier if this is a separate method ?

Reply:
Ok done. @steverendell